### DISCUSSION TITLE
raise_errors for yf.download similar to raise_errors for yf.Ticker(symbol).history

### DISCUSSION THREAD

**Discussion by abrhoda** [ranaroussi/yfinance#1832] [Q&A]
Hey all, I am using yfinance to run an analysis on a basket of symbols daily after market close. I am scraping wikipedia for all s&p 500 symbols and make a list out of them and then pass that list into `yf.download` function. Occasionally, I see the download function fail to download data for a few of the symbols in the list. Is there a way to raise errors similar to the `yf.Ticker(symbol).history(...)` method?

Relevant code snippet where I'm using the download function:
```
if (datetime.utcnow().time() < time(21, 00)) and (datetime.utcnow().time() > time(14, 30)):
    end_date = datetime.now() - timedelta(days=1)
    symbols_data = yf.download(config.symbols, start=config.data_start_date, end=end_date)
else:
    symbols_data = yf.download(config.symbols, start=config.data_start_date)
```

I have noticed that every so often, a few of the symbols will fail to get data from the api for any sort of reason and the script needs to be rerun and it will succeed. Example of running and getting the failed and then rerunning and seeing a success from earlier today:

![image](https://github.com/ranaroussi/yfinance/assets/30082436/f9da60f4-b0b4-41ce-b452-797ad75c03a7)

Looking through the code, I see that the [download](https://github.com/ranaroussi/yfinance/blob/176c3d628bd895cc06bf71d6af3b339be6d3413a/yfinance/tickers.py#L67) function calls the `multi.download` function which ultimately calls `Ticker(symbol).history(raise_errors=True, ...)` for each symbol in the list. `multi.download` aggregates the errors into `shared._ERRORS` and ends up logging via `logger.error(...)`. The errors are then never thrown as an exception and I don't see a flag to set for this similar to how `Ticker(symbol).history(...)` does.

A few questions:
1. Is there a way to raise errors similar to the `Ticker(symbol).history(...)` function so that I can exit the script with a non-zero exit code to signal to my cronjob to retry?
2. If not, is there a reason to not have this for the download function similar to the history method?
3. Also if not, is it better for me to manually be calling `yf.Ticker(symbol).history(raise_errors=True, ...)` for all symbols in the list instead of using the `yf.download` function in order to be able to see errors?

**Comment by ValueRaider** [comment]
Looks like a simple oversight, no reason `download` can't also have `raise_errors`. But some thought will be needed if multiple errors occurred - raise them all? just raise one? Maybe `raise_errors` argument allows choice?

Can see what "all" would look like with `yf.enable_debug_mode()` at top of script.

**Reply by abrhoda** [reply]
Sure, here's a screenshot of what the error looks like when it happens with debug logging enabled: 
![image](https://github.com/ranaroussi/yfinance/assets/30082436/fa92ecdd-2ee3-4f45-86c6-da86b3e394e9)

On line 46 of the screenshot, there is a list of symbols (just QRVO in this case) that is logged as error but not then raised to the caller. A few times I have run into this, this list is properly including more than one symbol like:
```
 ['QRVO', 'SYMBOL2']: Exception('%ticker%: No price data found, symbol may be delisted (1d 2023-01-16 -> 2024-01-16)
 ```
So it seems that it is aggregating the symbols and messages into a single log statement. I would lean towards a single exception being raised with the `list_of_symbols -> message` message similar to the log message. Or perhaps follow a .net style of aggregate exceptions and throw a AggregateException with field that is a list of the individual exceptions that occurred?

Diving into the debug log statements, you can also see on line 18 that the http request was attempted but no response came back? The script then moved onto the next symbol, NSC, without issue.

**Reply by ValueRaider** [reply]
I meant multiple *different* errors - rare but possible.

**Reply by abrhoda** [reply]
For the situation where there are multiple different errors and therefore multiple different exceptions, could a single `AggregateException` be raised that contains all the individual exceptions that were raised? If this is not the python way of doing things, please let me know as I don't work with python too often. Some rough code could look like: 

```
#! /usr/bin/env python

class AggregateException(Exception):
    def __init__(self, exceptions):
        super().__init__(f'exceptions: {exceptions}')
        self.exceptions = exceptions
    
    def __reduce__(self):
        return (AggregateExceptiion, (self.exceptions))

# If typed exceptions are wanted
class HttpClientException(Exception):
   pass

class HttpServerException(Exception):
   pass

if __name__ == '__main__':

    # plain exception example without unique types
    exception_1 = Exception("[SYMBOL]: No price data found, symbol may be delisted (1d 2023-01-17 -> 2024-01-17")
    exception_2 = Exception("[SYMBOL2, SYMBOL3]: Some other exception message that could happen.")
    exc_list = []
    exc_list.append(exception_1)
    exc_list.append(exception_2)
    aggregate_exception = AggregateException(exc_list)
    print(aggregate_exception)

    # typed exception example
    http_client_exception = HttpClientException("log client exception message")
    http_server_exception = HttpServerException("log server exception message")
    typed_exc_list = []
    typed_exc_list.append(http_client_exception)
    typed_exc_list.append(http_server_exception)
    typed_aggregate_exception = AggregateException(typed_exc_list)
    print(typed_aggregate_exception)
```

The AggregateException could then be raised and the caller could choose to look inside at the individual exceptions that were raised. I was thinking that the type exceptions would help with inspecting the individual exceptions to see if retrying (like in my case) with a HttpServerException would solve the case or if the symbol doesn't exist (such as a 404 http reponse code) and the type is HttpClientException so the caller knows not to retry. But with the untyped version just using the `Exception` class instead of `HttpClientException` and `HttpServerException`, the caller could also just iterate through the exceptions in the aggregate and check what the message is on the exception in order to decide if a retry would solve the issue or not.

Like I said, I don't work with python too much so let me know if there's a more conventional way to do this in python!

**Reply by ValueRaider** [reply]
I know nothing about aggregating exceptions. I'm fine with simply picking one to raise. Problem with raising all are the stack traces - one is enough to fill screen, never mind all. But which one?

**Reply by abrhoda** [reply]
I agree. Raising many exceptions, each with their own stacktraces, would get really messy from the terminal if you are just executing a script and they get this error. An aggregate exception is just an exception that contains many exceptions. Probably not what we're looking for here due what you brought up because the individual exceptions that the aggregate exception would retain their stacktracse and create a potentially very long message in the terminal. 

I will have some time this weekend so I'll create a fork and look into getting this working for raise_errors param on the multi downloads.

**Comment by abrhoda** [comment]
Finally got some time to look into this so following up here! I added the the `raise_errors` option to the multi download function and just changed the `shared._TRACEBACKS` dict to be the traceback object instead of only the string formatted version. This means in the download function, once all exceptions that occurred have been put into the dicts, the new `raise_errors` option can be used to either raise one (maybe the only) exception along with its original traceback to the caller or just logged (current behavior). I noticed there aren't unit tests for the `multi.py` file where the changes were made. I that something I should add? Attached is a quick screen shot from my terminal window where I ran a script with the `raise_errors` option on the download function set to 3 different values (False, defaulting to False, and True) along with a print statement saying what the case was handling. Is there any kind of testing that I should looking into outside of these manual test cases? I didn't see anything in the contribution guide about testing steps.

![image](https://github.com/ranaroussi/yfinance/assets/30082436/d940cb3e-420c-44fd-b091-768ab4072d76)

**Reply by ValueRaider** [reply]
> I should add?

At least one test that will fail if someone else breaks your code is a good idea, but "should" is up to you.

Are you randomly picking a stack trace to print? I prefer deterministic, easier to debug.

**Reply by abrhoda** [reply]
Sure. I'll go ahead and add a `multi.py` under the tests dir and test in there.

I am not just picking a random stack trace to print when multiple exceptions are thrown. They are put under `shared._ERRORS{}` and `shared._TRACEBACKS{}` and, if the `raise_errors` option in False (the defualt as well), then errors and tracebacks are printed as before. If `raise_errors` is true, I took the approach we discussed and am re-raising 1 exception with its original traceback to the caller. The exception that is raised is the first exception in the `shared._ERRORS` dict by just calling `ticker, err = list(shared._ERRORS.items())[0]`. Due to CPython ordering dicts since 3.6 (implementation detail) and officially in 3.7+, this will always return the first item added to the dict - from my understanding. So whichever the first exception that raised in the one that is re-raised to the caller. This is the best way I could think of to pick which is re-raised but open to any idea about a better/more deterministic way? Maybe the alphabetical first? That would require an additional sort though.

**Reply by ValueRaider** [reply]
Screenshot said "randomly picked", that's why I thought random.

As long as it's deterministic is fine. At some point though, it's better to submit the pull request instead of describing code.

**Reply by getorca** [reply]
any update on this? 

I don't follow the aggregating errors as I haven't dug through the code, but shouldn't it raise the first exception?

**Reply by ValueRaider** [reply]
#1084

**Comment by detlefm** [comment]
I also looked into the issue a few days ago.
I found it particularly problematic that I had errors as output in the console. Since nothing has happened here for some time, I have developed a solution outside of the core code and am offering it to everyone affected.   
You could certainly do it more elegantly and briefly with 'partial', but I couldn't think of a better way to keep the parameter list.

```
from typing import Union
import yfinance as yf
import pandas as pd
from yfinance.exceptions import YFException
import logging
from functools import wraps


# Custom handler that stores messages in a list
class ListHandler(logging.Handler):
    def __init__(self, log_list):
        super().__init__()
        self.log_list = log_list

    def emit(self, record):
        log_entry = self.format(record)
        self.log_list.append(log_entry)

# Decorator, raises error if any error logs happend
# depending on the raise_errors flag
# if raise_errors == False no exception is raised 
# but the log_lst contains all logged errors
def raise_if_log(log_lst:list,raise_errors:bool=False):
    def decorator(func):
        @wraps(func) 
        def wrapper(*args, **kwargs):
            nonlocal log_lst
            if log_lst == None:
                log_lst = []
            else:
                del log_lst[:]
            # create a separat logger
            logger = logging.getLogger("yfinance")
            logger.setLevel(logging.ERROR) 
            # Add a user handler
            handler = ListHandler(log_lst)
            formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')
            handler.setFormatter(formatter)
            logger.addHandler(handler)
            result = func(*args, **kwargs)
            if log_lst and raise_errors: # we have errors
                raise YFException(log_lst)
            return result
        return wrapper
    return decorator


# global variable to hold the log entries
loglst = []

# I want a parameter list for auto completion in vscode

@raise_if_log(log_lst=loglst,raise_errors=True)
def yf_download_raise(tickers, start=None, end=None, actions=False, threads=True,
             ignore_tz=None, group_by='column', auto_adjust=True, back_adjust=False,
             repair=False, keepna=False, progress=True, period="max", interval="1d",
             prepost=False, proxy=None, rounding=False, timeout=10, session=None,
             multi_level_index=True) -> Union[pd.DataFrame, None]:
     kwargs = locals()
     return yf.download(**kwargs)

# Two wrapper functions are neccessary because decorator parameters are static
@raise_if_log(log_lst=loglst,raise_errors=False)
def yf_download_noraise(tickers, start=None, end=None, actions=False, threads=True,
             ignore_tz=None, group_by='column', auto_adjust=True, back_adjust=False,
             repair=False, keepna=False, progress=True, period="max", interval="1d",
             prepost=False, proxy=None, rounding=False, timeout=10, session=None,
             multi_level_index=True) -> Union[pd.DataFrame, None]:
     kwargs = locals()
     return yf.download(**kwargs)


# test implementation 1
def test_raise(tickers, start=None, end=None, actions=False, threads=True,
             ignore_tz=None, group_by='column', auto_adjust=True, back_adjust=False,
             repair=False, keepna=False, progress=True, period="max", interval="1d",
             prepost=False, proxy=None, rounding=False, timeout=10, session=None,
             multi_level_index=True) -> Union[pd.DataFrame, None]:
    params = locals()
    try:
       df = yf_download_raise(**params) 
    except YFException as yfex:
        print('YFException')
        print(str(yfex))
    except Exception as ex:
        print('Generic Exception')
        print(str(ex))
    else:
        print('show dataframe')
        print(df)

# test implementation 2
def test_noraise(tickers, start=None, end=None, actions=False, threads=True,
             ignore_tz=None, group_by='column', auto_adjust=True, back_adjust=False,
             repair=False, keepna=False, progress=True, period="max", interval="1d",
             prepost=False, proxy=None, rounding=False, timeout=10, session=None,
             multi_level_index=True) -> Union[pd.DataFrame, None]:
    params = locals()
    df = yf_download_noraise(**params)
    if df.empty:
        print('Empty dataframe')
        for t in loglst:
            print(t)
    else:
        print(df)


print('start call with raise_errors, wrong symbol')
test_raise(tickers=['UNSINN'],period='5d',progress=False)
print('end call with raise_errors, wrong symbol')
print('-'*60,end="\n\n")
print('start call with raise_errors, right symbol')
test_raise(tickers=['MSF.DE'],period='5d',progress=False)
print('end with raise_errors, right symbol')
print('-'*60,end="\n\n")
print('start call with **not** raise_errors, wrong symbol')
test_noraise(tickers=['UNSINN'],period='5d',progress=False)
print('end call with **not** raise_errors, wrong symbol')
print('-'*60,end="\n\n")
print('call with **not** raise_errors, right symbol')
test_noraise(tickers=['MSF.DE'],period='5d',progress=False)
print('call with **not** raise_errors, wrong symbol')
print('-'*60,end="\n\n")



"""
Test output:
start call with raise_errors, wrong symbol
YFException
['2025-01-25 14:43:19,223 - ERROR - \n1 Failed download:', '2025-01-25 14:43:19,223 - ERROR - [\'UNSINN\']: YFPricesMissingError(\'$%ticker%: possibly delisted; no price data found  (period=5d) (Yahoo error = "No data found, symbol may be delisted")\')']
end call with raise_errors, wrong symbol
------------------------------------------------------------

start call with raise_errors, right symbol
show dataframe
Price            Close        High         Low        Open Volume
Ticker          MSF.DE      MSF.DE      MSF.DE      MSF.DE MSF.DE
Date
2025-01-20  413.799988  418.600006  412.649994  417.950012  30166
2025-01-21  411.399994  417.149994  410.000000  414.850006  23759
2025-01-22  425.700012  425.700012  414.649994  417.700012  32804
2025-01-23  426.149994  428.100006  424.149994  428.100006  46728
2025-01-24  424.299988  426.750000  421.100006  426.200012      0
end with raise_errors, right symbol
------------------------------------------------------------

start call with **not** raise_errors, wrong symbol
Empty dataframe
2025-01-25 14:43:20,856 - ERROR - 
1 Failed download:
2025-01-25 14:43:20,856 - ERROR -
1 Failed download:
2025-01-25 14:43:20,856 - ERROR -
1 Failed download:
2025-01-25 14:43:20,856 - ERROR - ['UNSINN']: YFPricesMissingError('$%ticker%: possibly delisted; no price data found  (period=5d) (Yahoo error = "No data found, symbol may be delisted")')
2025-01-25 14:43:20,856 - ERROR - ['UNSINN']: YFPricesMissingError('$%ticker%: possibly delisted; no price data found  (period=5d) (Yahoo error = "No data found, symbol may be delisted")')
2025-01-25 14:43:20,856 - ERROR - ['UNSINN']: YFPricesMissingError('$%ticker%: possibly delisted; no price data found  (period=5d) (Yahoo error = "No data found, symbol may be delisted")')
end call with **not** raise_errors, wrong symbol
------------------------------------------------------------

call with **not** raise_errors, right symbol
Price            Close        High         Low        Open Volume
Ticker          MSF.DE      MSF.DE      MSF.DE      MSF.DE MSF.DE
Date
2025-01-20  413.799988  418.600006  412.649994  417.950012  30166
2025-01-21  411.399994  417.149994  410.000000  414.850006  23759
2025-01-22  425.700012  425.700012  414.649994  417.700012  32804
2025-01-23  426.149994  428.100006  424.149994  428.100006  46728
2025-01-24  424.299988  426.750000  421.100006  426.200012      0
call with **not** raise_errors, right symbol
------------------------------------------------------------
"""
```

**Reply by ValueRaider** [reply]
#1084

**Reply by detlefm** [reply]
Thanks for the tip. But I had and have no intention of changing or improving anything inside the yfinance project.   
I just wanted to show a way using normal Python for users who would like an exception instead of a series of error logs.   
I don't know where such a change would make sense (if at all). 
Of course, it would be more elegant if the two calls that fetch the price information had the same API. But it is a library that has grown over time and must adapt to the changes in the Yahoo API and that is certainly enough work.   
If I come across something that I think would be an improvement, I will be happy to use this process.
